### PR TITLE
feat(server): Lobby server improvements

### PR DESCRIPTION
**Pull request by delucis**
Addresses some of the issues raised by #429 

1. Creates seat credentials when a player joins a game, deleting them when they leave, rather than instantiating them all once when a game is created. This stops credentials leaking as described by @flamecoals.

2. Enables some basic customisation of authentication by adding `generateCredentials` and `authenticateCredentials` options to the Lobby server.

    If provided, `generateCredentials` can be called in place of `shortid` and receives the Koa `ctx` object, so request headers etc. can be used for credential generation. This can be an async function if needed.

    `authenticateCredentials` receives a simplified signature of `(credentials, playerMetadata)` when called from the server master (compared to `{ action, gameMetadata, gameID, playerID }` used internally). In non-synchronous masters, `authenticateCredentials` can also be an async function to support third-party API calls.

There should be no breaking changes here. I changed the signature of the `SocketIO` server transport, but as far as I can tell, that wasn’t exposed externally in any way.

### CODE REVIEW

File: docs/documentation/api/Server.md [hunk 1]
```diff
@@ -24,6 +24,10 @@ A config object with the following options:
 
 3. `transport` (_object_): the transport implementation.
    If not provided, socket.io is used.
+   
+4. `generateCredentials` (_function_): an optional function that returns player credentials to store in the game metadata and validate against. If not specified, the Lobby’s `uuid` implementation will be used.
+
+5. `authenticateCredentials` (_function_): an optional custom function that tests if a player’s move is made with the correct credentials when using the default socket.io transport implementation.
```
Comment (nicolodavis) [inline docs/documentation/api/Server.md]: Let's call this "an optional function" (remove the word "custom") to be consistent with the line above.

File: docs/documentation/api/Server.md [hunk 2]
```diff
@@ -24,6 +24,10 @@ A config object with the following options:
 
 3. `transport` (_object_): the transport implementation.
    If not provided, socket.io is used.
+   
+4. `generateCredentials` (_function_): an optional function that returns player credentials to store in the game metadata and validate against. If not specified, the Lobby’s `uuid` implementation will be used.
```
Comment (nicolodavis) [inline docs/documentation/api/Server.md]: I feel like there is opportunity to merge `generateCredentials` and `uuid` into one thing.  Is there value in them being separate?

Comment (delucis) [reply docs/documentation/api/Server.md]: `uuid` is currently used to generate both player credentials and game IDs, which seem like quite separate things to me. That was fine for a context-less random generator, but I’m passing Koa’s `ctx` to `generateCredentials`, so that changes things a bit.

I also chose to have `generateCredentials` fall back first to `lobbyConfig.uuid` and then to `shortid.generate` as that preserves current behaviour, but falling back to `shortid` directly would probably make more sense. (The `lobbyConfig.uuid` option was misdocumented as `shortid` before, so I’m guessing there probably wouldn’t be too many people impacted if that changed.)

File: docs/documentation/api/Server.md (hunk 1 above)
Comment (delucis) [reply docs/documentation/api/Server.md]: Resolved in d07c8b2

File: docs/documentation/api/Server.md (hunk 2 above)
Comment (nicolodavis) [reply docs/documentation/api/Server.md]: Sorry if I missed some context in the original issue discussing this, but what's the use-case for having a customized function to generate credentials that isn't just invoking something like `uuid`?

**Comment by flamecoals** [comment]
I think previously the default unique ID was very long, so it made bad
links. I asked to change to a shorter one, which at the time we decided to
open up for customization. My only requirement is for it to be short and
preferably URL friendly

File: docs/documentation/api/Server.md (hunk 2 above)
Comment (delucis) [reply docs/documentation/api/Server.md]: The example I gave in the docs is of calling into an authentication service.

For example, Firebase Auth allows you to get an ID token for a logged in user on the client. You could then send that in the join game request, have `generateCredentials` decode the token, and return the decoded user ID (which is public) to be stored in the metadata. Then when making a move, the client could send along a similar (potentially refreshed) token and the `authenticateCredentials` method could again decode that token, and check if it matches the user ID stored in the metadata.

That could provide a way for users to authenticate that doesn’t rely on session storage.

**Comment by delucis** [comment]
I hadn’t thought about URLs. I have seen some people providing links to share for each seat when a game is created. I guess these changes would remove the credential leaking, but might prevent people from doing that kind of thing?

**Comment by flamecoals** [comment]
For freeboardgame.org, if I remember correctly, we share only the room id
and the server assigns the seat on the fly. But I need to review it to be
sure

File: docs/documentation/api/Server.md (hunk 2 above)
Comment (nicolodavis) [reply docs/documentation/api/Server.md]: Makes sense.  Thanks for clarifying!

**Comment by nicolodavis** [comment]
> I hadn’t thought about URLs. I have seen some people providing links to share for each seat when a game is created. I guess these changes would remove the credential leaking, but might prevent people from doing that kind of thing?

Why would it prevent people from creating links to join specific seats?  If `generateCredentials` returns something that is URL friendly, it should still be possible, no?

**Comment by delucis** [comment]
> If `generateCredentials` returns something that is URL friendly, it should still be possible, no?

Absolutely, but the credentials would have to be created one-by-one for each seat via the `join` endpoint (because they’re created on join, rather than on create game). That said, at the moment, you can only get a seat’s credentials by joining it, so presumably people are already “joining” several times to get the credentials and then displaying them all to a user.

**Comment by nicolodavis** [comment]
@delucis Yeah, we'll have to figure out what's the best way to support joining specific seats via a URL (I think that's a nice feature to have).  That's beyond the scope of this PR of course, so let's get this in first.

File: src/server/api.ts [hunk 3]
```diff
@@ -170,7 +180,8 @@ export const addApiToServer = ({ app, db, games, lobbyConfig }) => {
     }
 
     gameMetadata.players[playerID].name = playerName;
-    const playerCredentials = gameMetadata.players[playerID].credentials;
+    const playerCredentials = await lobbyConfig.generateCredentials(ctx);
```
Comment (nicolodavis) [inline src/server/api.ts]: Should we also pass `playerID` to `generateCredentials` (in case it wants to do something player specific)?

Comment (delucis) [reply src/server/api.ts]: Potentially. It is in theory already available as `ctx.request.body.playerID`, but there’s probably an argument for some handy aliases. Anything else you can think of that might be useful?

Comment (nicolodavis) [reply src/server/api.ts]: That's the only one that comes to mind at the moment.

Comment (nicolodavis) [reply src/server/api.ts]: I'm also happy leaving it out at the moment until we see a use case for it.  I'll let you make the call.

**Comment by nicolodavis** [comment]
@delucis What's the next step here?  Just double-checking that this is not waiting on me.

**Comment by delucis** [comment]
Sorry, dropped the ball on this one. It’s basically ready.

The one thing I’ve been indecisive on is the signature of the new `generateCredentials` and `authenticateCredentials` methods. The first just gets the raw Koa `ctx`, while the latter gets `credentials` from the player’s action payload and `playerMetadata`, which is the stored metadata for the player ID making the move. In other words, the first is very general (to allow use of custom headers in the `join` request), while the second is locked to some specific aliases (to hide the implementation complexity).

I kind of don’t like that these aren’t balanced and worry if they’re too specific to my experience. I forked a version of the server and integrated with Firebase Auth, but am not sure how helpful this `generate`/`authenticate` pattern would be for other services, because I don’t have experience using them.

In other words, I think it’s good to go and the fix for credential leaking is a definite improvement, I’m just fretting about details 😃

**Comment by nicolodavis** [comment]
That's fine.  We can change the signatures if we see a future use-case that would benefit from something a bit different.

File: src/server/index.ts [hunk 4]
```diff
@@ -33,14 +33,32 @@ export const createServerRunConfig = (portOrConfig?: any, callback?: any) => {
   return config;
 };
 
+/**
+ * Wrap a user-provided auth function to simplify external API
+ * @param  {function} fn The authentication function to wrap
+ * @return {function} Wrapped function for use by master
+ */
+const wrapAuthFn = fn => ({ action, gameMetadata, playerID }) =>
+  fn(action.payload.credentials, gameMetadata[playerID]);
```
Comment (nicolodavis) [inline src/server/index.ts]: Could you add a test that covers this line?

Comment (delucis) [reply src/server/index.ts]: Sure! Might take a moment for me to get to it, but I will.

Comment (nicolodavis) [reply src/server/index.ts]: No hurry. Thanks!
